x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java

```diff
@@ -129,7 +129,7 @@ ResolvedIndices resolveIndicesAndAliases(String action, IndicesRequest indicesRe
             final boolean replaceWildcards = indicesOptions.expandWildcardsOpen() || indicesOptions.expandWildcardsClosed();
 
             // check for all and return list of authorized indices
-            if (IndexNameExpressionResolver.isAllIndices(indicesList(indicesRequest.indices()))) {
+            if (isAllIndices(indicesRequest.indices())) {
```

Comment:
For ~10K indices, this change makes the method about 1-2 ms faster (~12ms vs ~10ms). This mostly helps requests that are directly created for REST level requests, e.g. `SearchRequest`.

Currently, the index authorization for `SearchRequest` with ~10k indices takes about 22ms and its divided as:
* loadAuthorizedIndices - ~3ms
* buildIndicesAccessControl - ~7ms
* ResolveIndices (this method) - ~12ms

So a 2ms reduction is close to 10% performance gain.

Reply:
I thought about making the change in `IndexNameExpressionResolver.isAllIndices`. But it has a test that specifically make sure `false` is returned for `*` ... Not sure whether that is still necessary. Anyway, limiting the change to where it is used is safer.

Reply:
It is surprising to me that not converting an empty array to an empty list saves 2 ms, while building a List of 10k elements takes 10 ms.
Or instead of the array->list conversion, it is the fact that `*` is handled here rather than `indexAbstractionResolver.resolveIndexAbstractions`? They have the same code, so I don't understand it either.
I might be missing something.

Reply:
Sorry for the confusion. I should have given more context in my initial comment. The time saving is observed for a specific field caps query during benckmarking tests:

```
GET /*/_field_caps?fields=*&ignore_unavailable=true&allow_no_indices=false
```
The option `ignore_unavailable=true` is critical here because that means the following code in the else branch will execute:
https://github.com/elastic/elasticsearch/blob/324996ed3265d3046f4e63230e666f31275212bc/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java#L152-L156

If `authorizedIndices` is large, `replaced` is also large (because it is an expansion of `*`), this loop and collecting are where the performance differs.

Reply:
Should we replace that stream with something more efficient then?
Like:
```
 if (indicesOptions.ignoreUnavailable()) { 
     //out of all the explicit names (expanded from wildcards and original ones that were left untouched) 
     //remove all the ones that the current user is not authorized for and ignore them 
     List<String> filtered = null;
     for (ListIterator<String> itr = replaced.listIterator(); itr.hasNext();) {
          final String index = itr.next();
          if (authorizedIndices.contains(index) == false) {
               if (filtered == null) {
                   filtered = new ArrayList(replaced.size()-1);
                   filtered.addAll( replaced.subList(0, itr.previousIndex()) );
               }
          } else {
               if (filtered != null) {
                   filtered.add(index);
               }
          }
    }
    if (filtered != null) {
         replaced = fitered
    }
 } 
```